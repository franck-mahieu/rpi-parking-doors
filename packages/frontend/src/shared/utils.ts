const env = process.env.NODE_ENV;
const reactAppDevApiUrl = process.env.REACT_APP_DEV_API_URL;

export const getBaseUrl = (
  envParam: string | undefined,
  reactAppApiUrlParam: string | undefined,
) => {
  return envParam === 'development' ? reactAppApiUrlParam : '';
};

export const fetchUtils = async (uri: string, method: string, body?: any) => {
  const response = await fetch(`${getBaseUrl(env, reactAppDevApiUrl)}${uri}`, {
    method: method,
    headers: {
      'Content-Type': 'application/json',
    },
    credentials: env === 'development' ? 'include' : 'same-origin',
    body: body,
  });
  const result: any = await response.json();
  if (
    (response.status >= 200 && response.status < 300) ||
    (result?.changes && result.changes !== 0)
  ) {
    return result;
  }
  return Promise.reject(result);
};
