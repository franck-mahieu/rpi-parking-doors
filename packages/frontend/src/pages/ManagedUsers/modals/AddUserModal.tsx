import React, { Component } from 'react';
import M from 'materialize-css';
import 'materialize-css/dist/css/materialize.min.css';
import { fetchUtils } from '../../../shared/utils';
import './Modals.scss';

const {
  REACT_APP_LABEL_ADD_USER_BUTTON,
  REACT_APP_LABEL_LOGIN,
  REACT_APP_LABEL_PASSWORD,
  REACT_APP_LABEL_ROLES,
  REACT_APP_ADD_USER_SUCCESS_MESSAGE,
  REACT_APP_ADD_USER_ERROR_MESSAGE,
  REACT_APP_LABEL_BUTTON_CANCEL,
  REACT_APP_LABEL_BUTTON_VALIDATE,
  REACT_APP_ADD_USER_MODAL_TITLE,
} = process.env;

interface IAddUserModal {
  guid: string;
  getUsers: any;
}

interface IAddUserModalState {
  login?: string;
  password?: string;
  roles?: string;
  email?: string;
  expiration?: string;
}

class AddUserModal extends Component<IAddUserModal> {
  private Modal: any;
  state: IAddUserModalState;

  constructor(props: any) {
    super(props);
    this.state = {
      login: '',
      password: '',
      roles: 'user',
      email: '',
      expiration: '',
    };
  }

  componentDidMount() {
    const options = {
      onOpenStart: () => {
        this.setState({
          login: '',
          password: '',
          roles: 'user',
          email: '',
          expiration: '',
        });
      },
      inDuration: 250,
      outDuration: 250,
      opacity: 0.5,
      dismissible: true,
      startingTop: '4%',
      endingTop: '10%',
    };

    M.Modal.init(this.Modal, options);
    const selectElement = document.querySelectorAll('select');
    M.FormSelect.init(selectElement);
    const datePickerElement = document.querySelectorAll('.datepicker');
    M.Datepicker.init(datePickerElement, {
      minDate: new Date(),
      defaultDate: new Date(),
      autoClose: true,
      showClearBtn: true,
      onSelect: date => {
        // usefull to keep the date of the selected day in different timezone
        date.setHours(12, 0, 0);
        this.setState({
          expiration: date,
        });
      },
    });
  }

  onClickAddUser = async () => {
    const { login, password, roles, email, expiration } = this.state;
    try {
      await fetchUtils(
        `/api/users/add?guid=${this.props.guid}`,
        'PUT',
        JSON.stringify({ login, password, roles, email, expiration }),
      );
      M.toast({
        html: `${login} ${REACT_APP_ADD_USER_SUCCESS_MESSAGE}`,
        classes: 'green',
      });
      this.props.getUsers();
    } catch (e) {
      M.toast({
        html: `${login} ${REACT_APP_ADD_USER_ERROR_MESSAGE}`,
        classes: 'red',
      });
    }
  };

  onChangeHandler = (event: any) => {
    if (event?.target?.id)
      this.setState({
        [event?.target?.id]: event?.target?.value,
      });
  };

  render() {
    return (
      <span id={'user-modal'}>
        <button
          className="waves-effect waves-light btn modal-trigger add-user-button"
          data-target="add-user-modal"
          title={REACT_APP_LABEL_ADD_USER_BUTTON}
        >
          {REACT_APP_LABEL_ADD_USER_BUTTON}
        </button>

        <div
          ref={Modal => {
            this.Modal = Modal;
          }}
          id="add-user-modal"
          className="modal"
        >
          <div className="modal-content">
            <h4>{REACT_APP_ADD_USER_MODAL_TITLE}</h4>
            <div className="row">
              <div className="input-field col s12">
                <input
                  id="login"
                  type="text"
                  className="validate"
                  value={this.state.login}
                  onChange={this.onChangeHandler}
                />
                <label htmlFor="login">{REACT_APP_LABEL_LOGIN}</label>
              </div>
            </div>
            <div className="row">
              <div className="input-field col s12">
                <input
                  id="password"
                  type="text"
                  className="validate"
                  value={this.state.password}
                  onChange={this.onChangeHandler}
                />
                <label htmlFor="password">{REACT_APP_LABEL_PASSWORD}</label>
              </div>
            </div>
            <div className="row">
              <div className="input-field col s12">
                <select onChange={this.onChangeHandler} id={'roles'}>
                  <option value="" disabled>
                    Choose the roles
                  </option>
                  <option value={'user'}>user</option>
                  <option value={'user admin'}>admin</option>
                </select>
                <label>{REACT_APP_LABEL_ROLES}</label>
              </div>
            </div>

            <div className="row">
              <div className="input-field col s12">
                <input
                  id="email"
                  type="email"
                  className="validate"
                  value={this.state.email}
                  onChange={this.onChangeHandler}
                />
                <label htmlFor="email">Email</label>
              </div>
            </div>

            <div className="row">
              <div className="input-field col s12">
                <input id="expiration" type="text" className="datepicker" />
                <label htmlFor="expiration">Expiration</label>
              </div>
            </div>
          </div>
          <div className="modal-footer">
            <button className="modal-close waves-effect waves-red btn-flat cancel-button">
              {REACT_APP_LABEL_BUTTON_CANCEL}
            </button>
            <button
              id={'add-user-validate-button'}
              className="modal-close waves-effect waves-green btn-flat"
              onClick={async () => {
                await this.onClickAddUser();
              }}
            >
              {REACT_APP_LABEL_BUTTON_VALIDATE}
            </button>
          </div>
        </div>
      </span>
    );
  }
}

export default AddUserModal;
