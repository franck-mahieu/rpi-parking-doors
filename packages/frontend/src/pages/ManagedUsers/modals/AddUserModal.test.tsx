import React from 'react';
import renderer from 'react-test-renderer';
import AddUserModal from './AddUserModal';

import * as utils from '../../../shared/utils';
import M from 'materialize-css';

jest.mock('../../../shared/utils');

describe('AddUserModal', () => {
  let addUserModal: AddUserModal;
  beforeEach(() => {
    jest.resetModules();
    jest.restoreAllMocks();
    //@ts-ignore
    utils.fetchUtils.mockResolvedValue(true);
    addUserModal = new AddUserModal({ guid: 'guid' });
  });

  describe('render', () => {
    it('renders correctly', () => {
      const tree = renderer
        .create(<AddUserModal guid={'guid'} getUsers={jest.fn()} />)
        .toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('should call onClickAddUser on click on button', () => {
      const component = renderer.create(
        <AddUserModal guid={'guid'} getUsers={jest.fn()} />,
      );
      const componentInstance: any = component.getInstance();
      jest.spyOn(componentInstance, 'onClickAddUser').mockResolvedValue('');
      component.root
        .findByProps({ id: 'add-user-validate-button' })
        .props.onClick();
      expect(componentInstance.onClickAddUser).toHaveBeenCalled();
    });
  });

  describe('componentDidMount', () => {
    jest.mock('materialize-css');
    it('should call M.Modal.init with valid params', async () => {
      const modalInitMock = jest.fn().mockImplementation((modal, options) => {
        options.onOpenStart();
      });
      const formSelectInitMock = jest.fn();
      const datePickerInitMock = jest
        .fn()
        .mockImplementation((modal, options) => {
          options.onSelect(new Date(2021, 3, 9));
        });
      M.Modal.init = modalInitMock;
      M.FormSelect.init = formSelectInitMock;
      M.Datepicker.init = datePickerInitMock;
      const setStateMock = jest.fn();
      addUserModal.setState = setStateMock;
      addUserModal.componentDidMount();
      expect(modalInitMock).toHaveBeenCalledWith(undefined, {
        onOpenStart: expect.any(Function),
        inDuration: 250,
        outDuration: 250,
        opacity: 0.5,
        dismissible: true,
        startingTop: '4%',
        endingTop: '10%',
      });
      expect(setStateMock).toHaveBeenNthCalledWith(1, {
        email: '',
        expiration: '',
        login: '',
        password: '',
        roles: 'user',
      });
      expect(setStateMock).toHaveBeenNthCalledWith(2, {
        expiration: new Date(2021, 3, 9, 12, 0, 0),
      });
    });
  });

  describe('onClickAddUser', () => {
    let toastMock: any;
    let componentInstance: any;
    beforeEach(() => {
      toastMock = jest.fn();
      M.toast = toastMock;
      const component = renderer.create(
        <AddUserModal guid={'guid'} getUsers={jest.fn()} />,
      );
      componentInstance = component.getInstance();
      componentInstance.setState({
        login: 'login',
        password: 'password',
        roles: 'user',
        email: 'email',
        expiration: 'expiration',
      });
    });

    it('should call fetchUtils and call M.toast with green when fetch is in sucess', async () => {
      await componentInstance.onClickAddUser();
      expect(utils.fetchUtils).toHaveBeenCalledWith(
        '/api/users/add?guid=guid',
        'PUT',
        '{"login":"login","password":"password","roles":"user","email":"email","expiration":"expiration"}',
      );
      expect(toastMock).toBeCalledWith({
        html: 'login a été ajouté avec succès',
        classes: 'green',
      });
    });

    it('should call fetchUtils and call M.toast with red when fetch is in error', async () => {
      //@ts-ignore
      utils.fetchUtils.mockRejectedValue();
      await componentInstance.onClickAddUser();
      expect(utils.fetchUtils).toHaveBeenCalledWith(
        '/api/users/add?guid=guid',
        'PUT',
        '{"login":"login","password":"password","roles":"user","email":"email","expiration":"expiration"}',
      );
      expect(toastMock).toBeCalledWith({
        html: "login n'a pas été ajouté, une erreur s'est produite",
        classes: 'red',
      });
    });
  });

  describe('onChangeHandler', () => {
    it('should call setState with key and value received', () => {
      const setStateMock = jest.fn();
      addUserModal.setState = setStateMock;
      addUserModal.onChangeHandler({
        target: { id: 'id', value: 'value' },
      });
      expect(setStateMock).toHaveBeenCalledWith({ id: 'value' });
    });

    it('should not call setState when received empty event', () => {
      const setStateMock = jest.fn();
      addUserModal.setState = setStateMock;
      addUserModal.onChangeHandler({});
      expect(setStateMock).not.toHaveBeenCalled();
    });
  });
});
