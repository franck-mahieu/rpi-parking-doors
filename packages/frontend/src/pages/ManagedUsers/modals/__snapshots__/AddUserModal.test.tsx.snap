// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddUserModal render renders correctly 1`] = `
<span
  id="user-modal"
>
  <button
    className="waves-effect waves-light btn modal-trigger add-user-button"
    data-target="add-user-modal"
    title="Ajouter un utilisateur"
  >
    Ajouter un utilisateur
  </button>
  <div
    className="modal"
    id="add-user-modal"
  >
    <div
      className="modal-content"
    >
      <h4>
        Ajouter un utilisateur
      </h4>
      <div
        className="row"
      >
        <div
          className="input-field col s12"
        >
          <input
            className="validate"
            id="login"
            onChange={[Function]}
            type="text"
            value=""
          />
          <label
            htmlFor="login"
          >
            Identifiant
          </label>
        </div>
      </div>
      <div
        className="row"
      >
        <div
          className="input-field col s12"
        >
          <input
            autoComplete="password"
            className="validate"
            id="password"
            onChange={[Function]}
            type="text"
            value=""
          />
          <label>
            Mot de passe
          </label>
        </div>
      </div>
      <div
        className="row"
      >
        <div
          className="input-field col s12"
        >
          <select
            id="roles"
            onChange={[Function]}
          >
            <option
              disabled={true}
              value=""
            >
              Choose the roles
            </option>
            <option
              value="user"
            >
              user
            </option>
            <option
              value="user admin"
            >
              admin
            </option>
          </select>
          <label>
            RÃ´les
          </label>
        </div>
      </div>
      <div
        className="row"
      >
        <div
          className="input-field col s12"
        >
          <input
            className="validate"
            id="email"
            onChange={[Function]}
            type="email"
            value=""
          />
          <label
            htmlFor="email"
          >
            Email
          </label>
        </div>
      </div>
      <div
        className="row"
      >
        <div
          className="input-field col s12"
        >
          <input
            className="datepicker"
            id="expiration"
            type="text"
          />
          <label
            htmlFor="expiration"
          >
            Expiration
          </label>
        </div>
      </div>
    </div>
    <div
      className="modal-footer"
    >
      <button
        className="modal-close waves-effect waves-red btn-flat cancel-button"
      >
        Annuler
      </button>
      <button
        className="modal-close waves-effect waves-green btn-flat"
        id="add-user-validate-button"
        onClick={[Function]}
      >
        Valider
      </button>
    </div>
  </div>
</span>
`;
